### PR TITLE
DAOS-5692 tools: Convert daos frontend to Go

### PR DESCRIPTION
**Pull request by mjmac**
Create a daos_hdlr library to allow the existing
daos tool functionality to be called from a new Go
frontend for the tool, and add new CLI handling code
under src/control/cmd/daos

Implementing the frontend in Go will harmonize the daos
tool with the dmg tool and allow sharing of UI code.

Co-authored-by: Bruno Faccini <bruno.faccini@intel.com>

### CODE REVIEW

File: src/control/cmd/daos/acl.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+#include "property.h"
+
+void
+free_strings(char **str, size_t str_count)
+{
+	int i;
+
+	for (i = 0; i < str_count; i++)
+		D_FREE(str[i]);
+}
```
Comment (johannlombardi) [inline src/control/cmd/daos/acl.go:23]: Should be removed?

File: src/control/cmd/daos/acl.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
+//
+// (C) Copyright 2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package main
+
```
Comment (johannlombardi) [inline src/control/cmd/daos/acl.go:8]: Maybe a comment here to say that this file is for parsons & converting ACL input strings into DAOS ACL data structure.

File: src/control/cmd/daos/attribute.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
+//
+// (C) Copyright 2018-2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
```
Comment (johannlombardi) [inline src/control/cmd/daos/attribute.go:6]: Same comment here for container attributes.

File: src/control/cmd/daos/container.go [hunk 4]
```diff
@@ -0,0 +1,834 @@
+//
+// (C) Copyright 2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
```
Comment (johannlombardi) [inline src/control/cmd/daos/container.go:6]: All "daos cont" operation handlers? It would be great to help identifying what are helper functions vs real operation handler. We have "static" in C and we usually add a comment to say that this is the handler. Would be great to have some equivalent here.

File: src/control/cmd/daos/container.go [hunk 5]
```diff
@@ -0,0 +1,834 @@
...
+	Type        string         `long:"type" short:"t" description:"container type" choice:"POSIX" choice:"HDF5" default:"POSIX"`
+	Path        string         `long:"path" short:"d" description:"container namespace path"`
+	ChunkSize   chunkSize      `long:"chunk-size" short:"z" description:"container chunk size"`
+	ObjectClass objectClass    `long:"oclass" short:"o" description:"default object class"`
+	Properties  PropertiesFlag `long:"properties" description:"container properties"`
+	ACLFile     string         `long:"acl-file" short:"A" description:"input file containing ACL"`
+	User        string         `long:"user" short:"u" description:"user who will own the container (username@[domain])"`
+	Group       string         `long:"group" short:"g" description:"group who will own the container (group@[domain])"`
+}
+
```
Comment (johannlombardi) [inline src/control/cmd/daos/container.go:123]: That's the hander for cont create, right?

File: src/control/cmd/daos/container.go [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+/*
+#include <daos.h>
+
+#include "daos_hdlr.h"
+*/
+import "C"
+
```
Comment (johannlombardi) [inline src/control/cmd/daos/container.go:32]: ah, maybe that's what i was looking for. A comment here saying that this is the primary table with all cmd handler would help.

File: src/tests/ftest/security/cont_get_acl.py [hunk 7]
```diff
@@ -68,7 +68,7 @@ def test_get_acl_valid(self):
                     outfile=path_to_file)
                 test_errs.extend(
                     self.error_handling(
-                        self.daos_cmd.result, "File exists"))
+                        self.daos_cmd.result, "ile exists"))
```
Comment (johannlombardi) [inline src/tests/ftest/security/cont_get_acl.py:71]: typo?

**Comment by johannlombardi** [comment]
Just a heads up that more properties will be added by Liang in PR #5832. Is that going to generate any conflicts?

File: src/control/cmd/daos/acl.go (hunk 1 above)
Comment (mjmac) [reply src/control/cmd/daos/acl.go:23]: No, this is how cgo works. You can write standalone code in .c/.h files in the same directory as the go package, or you can write small helpers in a comment before the `import "C"` statement. This `free_strings()` helper is used down on line 50.

That having been said, I added this before I would up adding the cmd/daos/util.c file and I wouldn't be opposed to moving it there if it reduces confusion.

File: src/tests/ftest/security/cont_get_acl.py (hunk 7 above)
Comment (mjmac) [reply src/tests/ftest/security/cont_get_acl.py:71]: heh, no. The error string from C's errno is capitalized, but in Go it's not. This test just matches on the suffix so now it works with either implementation.

**Comment by mjmac** [comment]
> Liang

Hmm, it will require some updates to the new frontend, yes. One of the UX improvements I am experimenting with in this change is to allow for tab-completion of property name/values.

**Comment by mjmac** [comment]
@johannlombardi I added a README file in the latest commit: https://github.com/daos-stack/daos/pull/5876/commits/fba578332b70ba75777e55d257cd1160691dfe12?short_path=8bbd094#diff-8bbd094e5968be98d0a09cce4fff598fcd85cf455e7fc32ddcbdd2b45213064f

Please take a look at let me know if that helps to answer some of the questions you might have about how things are organized, etc.

**Comment by mjmac** [comment]
NLT issues are warnings; I think there may need to be an exception added. Checkpatch issues are bogus IMO. I've fixed the other long-line issues.

File: src/control/cmd/daos/README.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	if err := daosError(rc); err != nil {
+		return errors.Errorf("failed to scrub container %s",
+			cmd.Container())
+	}
+
+	return nil
+}
+```
+
+There are more advanced possibilities here using the go-flags package. For example, we could change the Level argument from a simple string into a custom type that implements the `flags.Completer` interface in order to allow for tab-completion of valid scrub level names.
```
Comment (johannlombardi) [inline src/control/cmd/daos/README.md:143]: Looks very detailed. Thanks.

**Review by johannlombardi — APPROVED** [review]

File: src/control/cmd/daos/container.go (hunk 5 above)
Comment (tanabarr) [reply src/control/cmd/daos/container.go:123]: yep

File: src/control/cmd/daos/container.go [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+func (cmd *existingContainerCmd) ContainerID() ContainerID {
+	if !cmd.ContFlag.Empty() {
+		return cmd.ContFlag
+	}
+	return cmd.Args.Container
+}
+
+func (cmd *existingContainerCmd) resolveAndConnect(ap *C.struct_cmd_args_s) (cleanFn func(), err error) {
+	switch {
+	case cmd.Path != "" && !(cmd.PoolID().Empty() && cmd.ContainerID().Empty()):
```
Comment (tanabarr) [inline src/control/cmd/daos/container.go:247]: nit: extract the empty && empty into a local bar before the switch to DRY

File: utils/node_local_test.py [hunk 10]
```diff
@@ -581,24 +581,36 @@ def start(self, clean=True):
...
 
+        # This map contains resolutions for various states, with each
+        # tuple containing an error message to be matched and a command
+        # to be used to resolve the error. For example, if the
+        # command `dmg storage format` receives an error containing
+        # the message "storage format invoked on a running system",
+        # the resolution is to erase the system in order to restart
+        # the control plane with a fresh SCM mountpoint.
         error_resolutions = {
+            'storage_format': ( # starting state, no error
```
Comment (tanabarr) [inline utils/node_local_test.py:598]: these changes look unrelated to the PR

**Review by tanabarr — APPROVED** [review]
skimmed through, looks like good work Mike

**Review by tanabarr — APPROVED** [review]

File: src/control/cmd/daos/main.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		outputJSON(interface{}, error) error
+		errorJSON(error) error
+	}
+
+	jsonOutputCmd struct {
+		wroteJSON      *atm.Bool
+		writer         io.Writer
+		shouldEmitJSON bool
+	}
+)
```
Comment (kjacque) [inline src/control/cmd/daos/main.go:39]: Going to share this between daos and dmg? Looks like it's identical

**Review by johannlombardi — APPROVED** [review]

File: src/control/server/server.go [hunk 12]
```diff
@@ -163,16 +164,23 @@ func (srv *server) createServices(ctx context.Context) error {
...
+	srv.cbLock.Lock()
 	srv.onShutdown = append(srv.onShutdown, fns...)
+	srv.cbLock.Unlock()
 }
 
 func (srv *server) shutdown() {
-	for _, fn := range srv.onShutdown {
+	srv.cbLock.Lock()
+	onShutdownCbs := srv.onShutdown
+	srv.cbLock.Unlock()
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/server.go:182]: Heh, I'd completely forgotten that I added this mutex to quiet the race detector warning on shutdown.

So, yes. Assigning a slice variable to another variable doesn't make a copy of the slice, it's just copying the pointer to the slice. I think that this is safe in this case because what we're guarding against is unsafe concurrent access to the `srv.onShutdown` field. When `srv.OnShutdown()` is called, a new slice is created using `append()` and the pointer to that slice is saved in `srv.onShutdown`.

The reason that I'm doing the assignment under the lock rather than the loop is that I was concerned that a callback could do something that also needs the lock and then we've got a deadlock situation.

[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/server/server.go:182]: Ah, got it. Thanks. This should resolve DAOS-7809 as well.

**Review by kjacque — APPROVED** [review]
OK with landing as-is.
